crates/bevy_ecs/macros/src/component.rs

```diff
@@ -66,12 +73,40 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
         .predicates
         .push(parse_quote! { Self: Send + Sync + 'static });
 
+    let requires = &attrs.requires;
+    let mut register_required = Vec::with_capacity(attrs.requires.iter().len());
+    let mut register_recursive_requires = Vec::with_capacity(attrs.requires.iter().len());
+    if let Some(requires) = requires {
+        for require in requires {
+            let ident = &require.ident;
+            register_recursive_requires.push(quote! {
+                <#ident as Component>::register_required_components(components, storages, required_components);
+            });
+            if let Some(func) = &require.func {
+                register_required.push(quote! {
+                    required_components.register(components, storages, || { let x: #ident = #func().into(); x });
+                });
+            } else {
+                register_required.push(quote! {
+                    required_components.register(components, storages, <#ident as Default>::default);
+                });
+            }
+        }
+    }
     let struct_name = &ast.ident;
     let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
 
     TokenStream::from(quote! {
         impl #impl_generics #bevy_ecs_path::component::Component for #struct_name #type_generics #where_clause {
             const STORAGE_TYPE: #bevy_ecs_path::component::StorageType = #storage;
+            fn register_required_components(
+                components: &mut #bevy_ecs_path::component::Components,
+                storages: &mut #bevy_ecs_path::storage::Storages,
+                required_components: &mut #bevy_ecs_path::component::RequiredComponents
+            ) {
+                #(#register_required)*
+                #(#register_recursive_requires)*
+            }
```

Comment:
Could this document the reasoning for ordering `register_required` and `register_recursive_requires`? It seems critical for ensuring that the correct constructor is used for components that are required by more than one other component.